Filter return type inferences by constraint applicability

Andarist opened the pull request:
fixes https://github.com/microsoft/TypeScript/issues/50719
fixes https://github.com/microsoft/TypeScript/issues/41461
fixes https://github.com/microsoft/TypeScript/issues/57905

Andarist committed:
Filter return type inferences by constraint applicability

Andarist committed:
Merge remote-tracking branch 'origin/main' into fix/filter-out-return-type-inference-by-constraint-applicability

weswigham commented on src/compiler/checker.ts:
```diff
@@ -26852,14 +26852,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                }
+                else {
+                    if (inference.priority! & InferencePriority.ReturnType) {
+                        inference.inferredType = filterContextualInferredType(inferredType, instantiatedConstraint) ?? (fallbackType && filterContextualInferredType(fallbackType, instantiatedConstraint)) ?? instantiatedConstraint;
```
We only need to bother with this filtering if the constraint check for the whole union fails, right? (Since if it passes, no filtering will occur, as every union member must also pass.) This can probably be *inside* the comparison branch below?

Andarist replied:
That’s right but it felt cleaner to have this in a separate branch. Shouldnt filterType already optimize for array/type allocations?

jakebailey commented:
@typescript-bot test it

gabritto commented:
This had a ton of breaks on the extended tests, so seems like not a good idea?

Andarist commented:
yep, might not be a good idea after all - I didn't yet look at those breaks though, at the very least I'd like to get some new test cases based on them so we could add them to `main`

Andarist committed:
Merge remote-tracking branch 'origin/main' into fix/filter-out-return-type-inference-by-constraint-applicability

Andarist commented:
Slimmed down repro based on `headlessui`:
```ts
type Query = (
  container: HTMLElement,
  ...args: any[]
) =>
  | Error
  | HTMLElement
  | HTMLElement[]
  | Promise<HTMLElement[]>
  | Promise<HTMLElement>
  | null;

interface Queries {
  [T: string]: Query;
}

type FindByText<T extends HTMLElement = HTMLElement> = (
  container: HTMLElement,
  text: string,
) => Promise<T>;

declare function findByLabelText<T extends HTMLElement = HTMLElement>(
  ...args: Parameters<FindByText<T>>
): ReturnType<FindByText<T>>;

const queries = {
  findByLabelText,
};

type MapQueries<Q extends Queries = typeof queries> = {
  [P in keyof Q]: Q[P];
};
```

I didn't quite expect a change in any code like this... I'll continue investigating this.

Andarist committed:
fixed a silly mistake

jakebailey commented:
@typescript-bot test it

jakebailey commented:
@typescript-bot test it

Andarist committed:
add extra test case

Andarist committed:
add an extra test case

Andarist commented:
@gabritto I made a very silly mistake initially. It was somewhat a coincidence that the tests I had passed 😅 This was fixed now and I added extra test cases. The PR is now in a good state and can be reviewed :)

Above we can see some missing implicit anys in webpack. I added a test case based on that, this is a clear improvement.

ehoogeveen-medweb commented:
Aside from the new return type branch, it seems like this also changes the `!inferredType` case to always use `instantiatedConstraint` instead of checking whether `fallbackType` satisfies the constraint. Is that an intended part of this change?

Andarist committed:
dont forget about the fallback type

Andarist commented:
No, that wasn't intentional. Thanks for pointing that out - I corrected the PR but I doubt it will make a difference for any of the extended tests.

Andarist committed:
Merge remote-tracking branch 'origin/main' into fix/filter-out-return-type-inference-by-constraint-applicability

Andarist committed:
add extra test case

Andarist commented:
@gabritto @weswigham friendly 🏓 :)

Andarist committed:
Merge remote-tracking branch 'origin/main' into fix/filter-out-return-type-inference-by-constraint-applicability

Andarist committed:
Merge remote-tracking branch 'origin/main' into fix/filter-out-return-type-inference-by-constraint-applicability

Andarist committed:
match ts-go changes

jakebailey reviewed (approved):
Cross comparing typescript-go, and this is matching.
